### PR TITLE
[refactor] Removed Backend.get_<data> methods

### PR DESCRIPTION
**Pull request by lmondada**
I am currently making the necessary changes to pytket-extensions

### CODE REVIEW

**Comment by lmondada** [comment]
We're putting this on hold until we have cleared how exactly we'd like to do this.

**Comment by lmondada** [comment]
I'm now going to update the extensions.

File: pytket/pytket/backends/backend.py [hunk 1]
```diff
@@ -487,13 +363,37 @@ def get_results(
...
-    def _process(
-        self, circuit: Circuit, **kwargs: KwargTypes
-    ) -> Tuple[BackendResult, ResultHandle]:
-        handle = self.process_circuit(circuit, **kwargs)  # type: ignore
-        result = self.get_result(handle, **kwargs)
-        self.pop_result(handle)
-        return result, handle
+    def run_circuit(self, circuit: Circuit, **kwargs: KwargTypes) -> BackendResult:
+        """
+        Submit circuit to Backend `self` and returns results
```
Comment (cqc-alec) [inline pytket/pytket/backends/backend.py:368]:
```suggestion
        Submits a circuit to the backend and returns results
```

File: pytket/pytket/backends/backend.py [hunk 2]
```diff
@@ -487,13 +363,37 @@ def get_results(
...
+
+        :param circuit: Circuit to be executed
+        :return: Result
+
+        This is a conveniance method equivalent to calling
+        `self.process_circuit` followed by `self.get_result`.
```
Comment (cqc-alec) [inline pytket/pytket/backends/backend.py:374]:
```suggestion
        :py:meth:`Backend.process_circuit` followed by :py:meth:`Backend.get_result`.
```

File: pytket/pytket/backends/backend.py [hunk 3]
```diff
@@ -487,13 +363,37 @@ def get_results(
...
+        Any keyword arguments are passed on to `process_circuit` and `get_result`.
+        """
+        return self.run_circuits([circuit], **kwargs)[0]
+
+    def run_circuits(
+        self, circuits: Iterable[Circuit], **kwargs: KwargTypes
+    ) -> List[BackendResult]:
+        """
+        Submit circuits to Backend `self` and returns results
```
Comment (cqc-alec) [inline pytket/pytket/backends/backend.py:383]:
```suggestion
        Submits circuits to the backend and returns results
```

File: pytket/pytket/backends/backend.py [hunk 4]
```diff
@@ -487,13 +363,37 @@ def get_results(
...
+
+        :param circuits: Iterable of Circuits to be executed
+        :return: List of results
+
+        This is a convenience method equivalent to calling
+        `self.process_circuits` followed by `self.get_results`.
```
Comment (cqc-alec) [inline pytket/pytket/backends/backend.py:389]:
```suggestion
        :py:meth:`Backend.process_circuits` followed by :py:meth:`Backend.get_results`.
```

File: pytket/pytket/backends/backend.py [hunk 5]
```diff
@@ -487,13 +363,37 @@ def get_results(
...
+        Any keyword arguments are passed on to `process_circuits` and `get_results`.
+        """
+        handles = self.process_circuits(circuits, **kwargs)  # type: ignore
```
Comment (cqc-alec) [inline pytket/pytket/backends/backend.py:392]: Why do we need `type: ignore` here?

**Comment by lmondada** [comment]
A draft `pytket-extensions` PR has been opened: https://github.com/CQCL/pytket-extensions/pull/202. When this is merged in, we can run the CI tests on the extensions.

**Review by cqc-alec — CHANGES_REQUESTED** [review]

File: pytket/pytket/backends/backend.py (hunk 5 above)
Comment (lmondada) [reply pytket/pytket/backends/backend.py]: Mypy is a bit fidly with `**kwargs`. I have found another fix without `type: ignore` though.

File: pytket/pytket/backends/backend.py (hunk 1 above)
Comment (lmondada) [reply pytket/pytket/backends/backend.py]: done

File: pytket/pytket/backends/backend.py (hunk 2 above)
Comment (lmondada) [reply pytket/pytket/backends/backend.py]: done

File: pytket/pytket/backends/backend.py (hunk 3 above)
Comment (lmondada) [reply pytket/pytket/backends/backend.py]: done

File: pytket/pytket/backends/backend.py (hunk 4 above)
Comment (lmondada) [reply pytket/pytket/backends/backend.py]: done

**Review by cqc-alec — APPROVED** [review]
Looks good to me; after merging we can make a candidate release to enable the extensions and manual/examples to be updated.

**Pull request merged**